File: Tuning/sample_code/04_alloc2d/src/fortran/main.f90
! Copyright 2024 Research Organization for Information Science and Technology
!--------------------------------------------------------------------------
!  Address mapping of 2D array in Fortran 90
!  Author:      Yukihiro Ota (yota@rist.or.jp)
!  Last update: 27 Oct. 2017
!--------------------------------------------------------------------------
program main
  implicit none

  integer,parameter :: nrow = 3
  integer,parameter :: ncol = 8
  integer :: i, j
  integer :: seed_size
#ifndef __ARCH_32__
  integer(8) :: adr_x 
#else
  integer :: adr_x
#endif 
  integer,allocatable :: seed(:)

  real(8) :: xtmp 
  real(8) :: x_s(nrow,ncol)
  real(8),allocatable :: x_d(:,:) 

  allocate ( x_d(1:nrow,1:ncol) )
  !------------------------------------------------------------------------
  ! initialization by uniform random number in [0,1) 
  !------------------------------------------------------------------------
  call random_seed (size=seed_size) 

  allocate( seed(seed_size) )
  
  call random_seed (get=seed) 

  call system_clock (count=i)
  do i=1,seed_size
    call system_clock (count=seed(i))
  enddo

  call random_seed (put=seed)

  do j=1,ncol
    do i=1,nrow
      call random_number (xtmp)
      x_s(i,j) = xtmp 
      x_d(i,j) = xtmp 
    enddo
  enddo
  !------------------------------------------------------------------------
  ! main calculation
  !------------------------------------------------------------------------
  write (6,'(A)') "================================================="
  write (6,'(A)') "Address mapping of 2D array in Fortran 90        "
  write (6,'(A)') "  Author:      Yukihiro Ota (yota@rist.or.jp)    "
  write (6,'(A)') "  Last update: 27 Oct 2017                       " 
  write (6,'(A)') "================================================="

  write (6,'(A)') "Static memory allocation (real(8)=8 bytes)" 
  write (6,'(A)') "x_s(i,j)"
  write (6,'(A)') "row     col     val         address (dec) "
  do j=1,ncol
    do i=1,nrow
      adr_x = loc (x_s(i,j))  !! loc is GNU extensions
      write (6, '(2I5, 1F11.4,1I20)') i, j, x_s(i,j), adr_x 
    enddo
  enddo

  write (6,'(A)') "Dynamical memory allocation (real(8)=8 bytes)" 
  write (6,'(A)') "x_d(i,j)"
  write (6,'(A)') "row     col     val         address (dec)"
  do j=1,ncol
    do i=1,nrow
      adr_x = loc (x_d(i,j))  !! loc is GNU extensions
      write (6, '(2I5, 1F11.4,1I20)') i, j, x_d(i,j), adr_x 
    enddo
  enddo

  ! Finalization
  deallocate ( seed ) ;

  deallocate ( x_d )  

  stop
end program main 

File: Tuning/sample_code/04_alloc2d/src/c/main.c
/* Copyright 2024 Research Organization for Information Science and Technology */
/*------------------------------------------------------------------------
  Address mapping of 2D array in C
  Author:      Yukihiro Ota (yota@rist.or.jp)
  Last update: 27 Oct 2017
------------------------------------------------------------------------*/
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int main ( int argc, char **argv ) 
{
  const int nrow = 3 ;  
  const int ncol = 8 ;  
  double   x_s[nrow][ncol] ; /* static        */
  double** x_d ;             /* dynamical (1) */
  double** x_dc_ptr ;        /* dynamical (2) */
  double*  x_dc ;

  /* Memory allocation                                                   */
  x_d = (double **)malloc ( sizeof(double *)*nrow ) ;
  for ( int i=0; i<nrow; ++i ) {
    x_d[i] = (double *)malloc ( sizeof(double)*ncol ) ;
  }

  x_dc_ptr = (double **)malloc ( sizeof(double *)*nrow ) ;
  x_dc = (double *)malloc ( sizeof(double)*nrow*ncol ) ; /* 1D array */
  for ( int i=0; i<nrow; ++i ) {
    x_dc_ptr[i] = x_dc + i*ncol ; /* pointer to 1D array per row */
  }
  /*---------------------------------------------------------------------*/
  /* Initialization of data with random number in [0,1)                   */
  /*---------------------------------------------------------------------*/
  {
    srand( (unsigned int)time(NULL) ) ;
    for ( int i=0; i<nrow; ++i ) {
      for ( int j=0; j<ncol; ++j ) {
        double xtmp = 2.0*(double )rand()/RAND_MAX - 1.0 ;
        x_s[i][j]  = xtmp ;
        x_d[i][j]  = xtmp ;
        x_dc_ptr[i][j] = xtmp ;
      }
    }
  }
  /*---------------------------------------------------------------------*/
  /* main calculation                                                    */
  /*---------------------------------------------------------------------*/
  printf ("======================================================\n");
  printf ("Address mapping of 2D array in C                      \n");
  printf ("   Author:      Yukihiro Ota (yota@rist.or.jp)        \n");
  printf ("   Last update: 27 Oct 2017                           \n");
  printf ("======================================================\n");

  printf ("Static memory allocation in (double=8bytes)\n") ;
  printf ("x_s[i][j]\n") ;
  printf ("row    col    val         address (hex) \n") ;
  for ( int i=0; i<nrow; ++i ) {
    for ( int j=0; j<ncol; ++j ) {
      printf ("%5d  %5d  %11.4f  %p\n", i, j, x_s[i][j], &x_s[i][j] ) ;
    }
  }

  printf ("Typical dynamical memory allocation (double=8bytes)\n") ;
  printf ("x_d[i][j]\n") ;
  printf ("row    col    val         address (hex) \n") ;
  for ( int i=0; i<nrow; ++i ) {
    for ( int j=0; j<ncol; ++j ) {
      printf ("%5d  %5d  %11.4f  %p\n", i, j, x_d[i][j], &x_d[i][j] ) ;
    }
  }

  printf ("Alternative Dynamical memory allocation (double=8bytes)\n") ;
  printf ("x_dc_ptr[i][j]\n") ;
  printf ("row    col    val         address (hex) \n") ;
  for ( int i=0; i<nrow; ++i ) {
    for ( int j=0; j<ncol; ++j ) {
      printf ("%5d  %5d  %11.4f  %p\n", i, j, x_dc_ptr[i][j], 
              &x_dc_ptr[i][j] ) ;
    }
  }
  printf ("x_dc[k] with k = j + i*(num of col.)\n") ;
  printf ("row    col    val         address (hex) \n") ;
  for ( int i=0; i<nrow; ++i ) {
    for ( int j=0; j<ncol; ++j ) {
      int k = j + i*ncol ;
      printf ("%5d  %5d  %11.4f  %p\n", i, j, x_dc[k], &x_dc[k] ) ;
    }
  }


  /* Finalization */
  for ( int i=0; i<nrow; ++i ) {
    free ( x_d[i] ) ;
  }
  free ( x_d ) ; 

  free ( x_dc_ptr ) ; 
  free ( x_dc ) ;

  return EXIT_SUCCESS ;
}
